### easyGWASCore

Core Framework of easyGWAS (http://easygwas.ethz.ch) for computing genome-wide association studies and meta-analysis. This is a C/C++ Framework with Python interfaces. The code includes several standard methods for performing GWAS, such as linear regression, logistic regression and popular linear mixed models (EMMAX (http://www.nature.com/ng/journal/v42/n4/abs/ng.548.html), FaSTLMM (http://www.nature.com/nmeth/journal/v8/n10/abs/nmeth.1681.html)) to also account for population stratification. In addition, the package contains code for the network guided multi-locus mapping method SConES (http://bioinformatics.oxfordjournals.org/content/29/13/i171.short).

### Tutorials

Tutorials can be found in the Wiki section of this repository: http://github.com/dominikgrimm/easyGWASCore/wiki

### License

Code by: Dominik Gerhard Grimm
Year: 2011-2016
Group: Machine Learning and Computational Biology Research Group
Insitute: Max Planck Institute for Intelligent Systems, Max Planck Institute for Developmental Biology and ETH ZÃ¼rich

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

